src/software/amazon/ion/impl/lite/IonIntLite.java

```diff
@@ -34,11 +35,18 @@
     static private final BigInteger LONG_MAX_VALUE =
         BigInteger.valueOf(Long.MAX_VALUE);
 
+    static private final BigInteger INT_MIN_VALUE =
+        BigInteger.valueOf(Integer.MIN_VALUE);
+
+    static private final BigInteger INT_MAX_VALUE =
+        BigInteger.valueOf(Integer.MAX_VALUE);
+
     private static final int HASH_SIGNATURE =
         IonType.INT.toString().hashCode();
 
     private long _long_value;
     private BigInteger _big_int_value;
+    private IntegerType _actual_type;
```

Comment:
Collapsing to `BigInteger` for `long` (for a common case) would be relatively expensive.  You could do it with a `Long` wrapper for this *union*, but that would make the `long` case be a pointer + integer.

You could make the determination of the type to be lazy and eliminate that field.